src/parser.cpp

```diff
@@ -1009,3 +1009,110 @@ Writer<std::unique_ptr<AST::AST>> Parser::parse_type_term() {
+	    {"Parse Error: Failed to parse type var"}};
+
+	if (handle_error(result, require(TokenTag::AT)))
+		return result;
+
+	auto token = require(TokenTag::IDENTIFIER);
+	if (handle_error(result, token))
+		return result;
+
+	auto t = std::make_unique<AST::TypeVar>();
+	t->m_token = token.m_result;
+
+	return make_writer<std::unique_ptr<AST::AST>>(std::move(t));
+}
+
+Writer<std::unique_ptr<AST::AST>> Parser::parse_type_var() {
+	Writer<std::unique_ptr<AST::AST>> result = {
+	    {"Parse Error: Failed to parse type function"}};
+
+	if (consume(TokenTag::KEYWORD_UNION)) {
+		if (handle_error(result, require(TokenTag::BRACE_OPEN)))
+			return result;
+
+		std::vector<std::unique_ptr<AST::AST>> constructors;
+		std::vector<std::unique_ptr<AST::AST>> types;
+		while(!consume(TokenTag::BRACE_CLOSE)) {
+			auto cons = parse_identifier();
+			if (handle_error(result, cons))
+				return result;
+			
+			if (handle_error(result, require(TokenTag::COLON)))
+				return result;
+
+			auto type = parse_type_term();
+			if (handle_error(result, type))
+				return result;
+
+			if (handle_error(result, require(TokenTag::SEMICOLON)))
+				return result;
+
+			constructors.push_back(std::move(cons.m_result));
+			types.push_back(std::move(type.m_result));
+		}
+
+		auto u = std::make_unique<AST::Union>();
+		u->m_constructors = std::move(constructors);
+		u->m_types = std::move(types);
+
+		return make_writer<std::unique_ptr<AST::AST>>(std::move(u));
+	} else if (consume(TokenTag::KEYWORD_TUPLE)) {
+		if (handle_error(result, require(TokenTag::BRACE_OPEN)))
+			return result;
+
+		std::vector<std::unique_ptr<AST::AST>> types;
+		while(!consume(TokenTag::BRACE_CLOSE)) {
+
+			auto type = parse_type_term();
+			if (handle_error(result, type))
+				return result;
+
+			if (handle_error(result, require(TokenTag::SEMICOLON)))
+				return result;
+
+			types.push_back(std::move(type.m_result));
+		}
+
+		auto t = std::make_unique<AST::Tuple>();
+		t->m_types = std::move(types);
+
+		return make_writer<std::unique_ptr<AST::AST>>(std::move(t));
+	} else if (consume(TokenTag::KEYWORD_STRUCT)) {
+		if (handle_error(result, require(TokenTag::BRACE_OPEN)))
+			return result;
+
+		std::vector<std::unique_ptr<AST::AST>> fields;
+		std::vector<std::unique_ptr<AST::AST>> types;
+		while(!consume(TokenTag::BRACE_CLOSE)) {
+			auto field = parse_identifier();
+			if (handle_error(result, field))
+				return result;
+			
+			if (handle_error(result, require(TokenTag::COLON)))
+				return result;
+
+			auto type = parse_type_term();
+			if (handle_error(result, type))
+				return result;
+
+			if (handle_error(result, require(TokenTag::SEMICOLON)))
+				return result;
+
+			fields.push_back(std::move(field.m_result));
+			types.push_back(std::move(type.m_result));
+		}
+
+		auto s = std::make_unique<AST::Struct>();
+		s->m_fields = std::move(fields);
+		s->m_types = std::move(types);
+
+		return make_writer<std::unique_ptr<AST::AST>>(std::move(s));
+	}
```

Comment:
There is some pretty blatant duplication here (struct and union body is exactly the same) can you do something about it?